analysis_templates/cms_minimal/__cf_module_name__/weight/example.py

```diff
@@ -0,0 +1,43 @@
+# coding: utf-8
+
+"""
+Example event weight producer.
+"""
+
+from columnflow.weight import WeightProducer, weight_producer
+from columnflow.util import maybe_import
+from columnflow.config_util import get_shifts_from_sources
+from columnflow.columnar_util import Route
+
+ak = maybe_import("awkward")
+np = maybe_import("numpy")
+
+
+@weight_producer(
+    # both produced columns and dependent shifts are defined in init below
+    # only run on mc
+    mc_only=True,
+)
+def example(self: WeightProducer, events: ak.Array, **kwargs) -> ak.Array:
+    # build the full event weight
+    weight = ak.Array(np.ones(len(events), dtype=np.float32))
+    for column in self.weight_columns:
+        weight = weight * Route(column).apply(events)
+
+    return weight
```

Comment:
I agree with you that this would be more consistent, but at the same time this would mean that users could use the `WeightProducer` to modify their event content (either by accident or on purpose) during the creation of histograms.

I personally would welcome this change, because more flexibility is always nice when you know what you're doing, but I'd also understand if we decide to not allow this.

Reply:
If we add this, we could also close https://github.com/columnflow/columnflow/issues/379, because then, people could add behaviour themselves during histogramming.

Reply:
@riga do you have something to add to this discussion? If not, I'd implement this change

Reply:
Indeed, modifying the event array would then become possible, which could provide additional pitfalls for users or be misused. OTOH the impact should be minimal and it also gives users more flexibility, as in the issue you cited. 

Taking this into consideration, I still think it would be preferable to add the `events` in order to keep the same return signature for all `TaskArrayFunction` objects, i.e. always return the events as the first element of the return tuple. Aside from being more consistent, this might also have implications with how caching is implemented, where IIRC we make the assumption that the event array is returned first.